pkg/models/shipment.go

```
@@ -148,6 +148,40 @@ func (s *Shipment) Accept() error {
 	return nil
 }
 
+// BeforeSave will run before each create/update of a Shipment and will attempt to determine
+// and set the associated TDL given the pickup and destination zip codes.
+func (s *Shipment) BeforeSave(tx *pop.Connection) error {
+	var rateArea, region string
+	var err error
+
+	// Look up TDL and set if we can determine source rate area, region, and COS.
+	if s.PickupAddress != nil {
+		rateArea, err = FetchRateAreaForZip5(tx, s.PickupAddress.PostalCode)
+		if err != nil {
+			return errors.Wrap(err, "Could not fetch rate area")
+		}
+	}
+	if s.DeliveryAddress != nil {
+		region, err = FetchRegionForZip5(tx, s.DeliveryAddress.PostalCode)
+		if err != nil {
+			return errors.Wrap(err, "Could not fetch region")
+		}
+	}
+	codeOfService := "D" // Always hard-coded for now.
+
+	if rateArea != "" && region != "" && codeOfService != "" {
```

I'm not sure what this check is for. It would only be true if our data for `FetchRateAreaForZip5` had blanks in the database, which I hope it doesn't. If it does, I'd think we'd want to return that error right after we fetched. Also, there is no `else`, so any errors are silently dropped.

---

Looking at the Shipment model/schema, the pickup address and delivery address are both optional.  Based on that, I made the assumption there are times we may be saving a Shipment record but not have all the necessary parts (pickup, delivery, CoS) required to associate a TDL -- and hence, that is a legitimate state and shouldn't be reported as an error condition.  So, the way I handled it is to set the rateArea and region to their zero values initially in Go, then only attempt lookups (and report errors) for rate area, region, and TDL if I had the appropriate source data.

That said, my assumptions may be off here.

---

Maybe that's part of the problem with my testing.

I believe the shipment model (like the PPM model) gets PATCHed on every page of the Wizard (or will, when we have multiple pages,) and so we want to be able to save incomplete shipment records.

That said, I don't think we're making multiple PATCHes to the shipment record yet, since the service member form is submitted in one page right now. In any case, I'd look at that use case (the service member form) and make sure this approach works in that context.

---

BeforeSave has been reworked to do the following:

1) If the shipment already has an associated TDL, do nothing.  If we have concerns that the shipment's TDL could get out of sync with the addresses, we can remove this check (with some possible small performance hit).

2) If we don't have a pickup address (it's optional in the model/db), then do nothing since we don't have one of the parts needed to determine the TDL.

3) Use the destination duty station to determine the pickup zip as opposed to the destination address (since the latter is optional, at least at this point in the process).

4) Create a new TDL record if the one we're looking for doesn't already exist.